### ISSUE TITLE
if (xx = x);  typeof(pointerof(xx))  is still nillable

### ISSUE THREAD

**Issue by oprypin** [crystal-lang/crystal#9698]
##### Semi-related example without pointers - good:
```crystal
x = (rand >= 0 ? 6 : nil)
if (xx = x)
  p! typeof(xx)  # => Int32
end
```
### **Problem**: taking a pointer of a for-sure non-nil variable still includes Nil in the union:
```crystal
x = (rand >= 0 ? 6 : nil)
if (xx = x)
  p! typeof(pointerof(xx))  # => Pointer(Int32 | Nil)
end
```
#### Workaround:
```crystal
x = (rand >= 0 ? 6 : nil)
if x
  xx = x
  p! typeof(pointerof(xx))  # => Pointer(Int32)
end
```
---
#### Why I ran into this / why this is a problem.
I was trying to bind a library which has something similar as part of its interface:
```crystal
lib Lib
  fun f(x : Int32*)
end
```
Then my wrapper ([naïve attempt](https://carc.in/#/r/9lb4)):
```crystal
def f(x : Int32?)
  Lib.f(pointerof(x))
end
# Error: argument 'x' of 'Lib#f' must be Pointer(Int32), not Pointer(Int32 | Nil)
```crystal
[Less naïve](url):
x = (rand >= 0 ? 6 : nil)
Lib.f(x ? pointerof(x) : Pointer(Int32).null)
```

**Comment by oprypin** [comment]
#### Why I ran into this / why this is a problem.
I was trying to bind a library which has something similar as part of its interface:
```c
void f (int* x) { if (x) printf("%d\\n", *x); }
```
```crystal
lib Lib
  fun f(x : Int32*)
end
```
Then my wrapper ([naïve attempt](https://carc.in/#/r/9lb4)):
```crystal
def f(x : Int32?)
  Lib.f(pointerof(x))
end
# Error: argument 'x' of 'Lib#f' must be Pointer(Int32), not Pointer(Int32 | Nil)
```
[Less naïve](https://carc.in/#/r/9lb5):
```crystal
x = (rand >= 0 ? 6 : nil)
Lib.f(x ? pointerof(x) : Pointer(Int32).null)
# Error: argument 'x' of 'Lib#f' must be Pointer(Int32), not (Pointer(Int32 | Nil) | Pointer(Int32))
```
Mhm, getting more complicated. [How about I force it!](https://carc.in/#/r/9l9s)
```crystal
x = (rand >= 0 ? 6 : nil)
Lib.f(x ? pointerof(x).as(Int32*) : Pointer(Int32).null)
# => 163
```
Mmmm what's that 163 now? Ah yes, the `typeid` of `Int32`.

Well I'm glad that I found out about this accidentally. Normally I might have just [tried the basic cases](https://carc.in/#/r/9lb8):
```crystal
def f(x : Int32?)
  Lib.f(x ? pointerof(x).as(Int32*) : Pointer(Int32).null)
end
f(6)  # 6
f(nil)  # (nothing)
```
This works fine! And would've silently exploded as soon as a user of this wrapper happened to pass a union of `Int32 | Nil`

**Comment by oprypin** [comment]
Of course, as per the main post, now I know how to [deal with this](https://carc.in/#/r/9laq):
```crystal
Lib.f(x && (xx = x; pointerof(xx)))
```
(this is just a rephrasing of the workaround).

But the thing is, I shouldn't have arrived at the horribly flawed solution; `x ? pointerof(x) : Pointer(Int32).null` should work, at least!

**Comment by oprypin** [comment]
#### What's really happening
Once you realize that this is wrong in *some* way:
```crystal
x = (rand >= 0 ? 6 : nil)
Lib.f(x ? pointerof(x).as(Int32*) : Pointer(Int32).null)
```
it's quite easy to see why. `x` really is of a union type, and the fact that we checked that it's not the `Nil` option doesn't magically change its memory representation as a union, that is `{typeid, data_union}`. Forcibly converting the pointer makes us look at the `typeid` field...

Now, apparently, Crystal's implementation is very aware of the limitations of the memory layout when dealing with code like this:
```crystal
x = (rand >= 0 ? 6 : nil)
if x
  pointerof(x)  # .is_a?(Pointer(Int32 | Nil))
end
```
Inside the `if`, even though both the user and the compiler know that `x` can only be `Int32`, it's still represented as a union in memory, and taking a pointer of the union, gives you a pointer typed as the union!
Unfortunately, even this doesn't help:
```crystal
if (xx = x)
  pointerof(xx)  # .is_a?(Pointer(Int32 | Nil))
end
```
-- I suppose it doesn't change anything really, `xx` is still copied as a union and only then confirmed to not be `Nil`.
The following, however, helps! We're dealing with `x` which is a union but known to be only of one type, so when copying it into `xx` Crystal doesn't need to make *that* one a union too. So finally we get our proper pointer.
```crystal
if x
  xx = x
  pointerof(xx)  # .is_a?(Pointer(Int32))
end
```
But the thing is, Crystal *could* satisfy this use case, in the end there's nothing preventing the implementation from automatically applying a workaround, actually something even more efficient, not involving a copy. Basically an equivalent of [**this**](https://carc.in/#/r/9lbc):
```crystal
Lib.f(x ? (pointerof(x).as(Void*) + 8).as(Int32*) : Pointer(Int32).null)  # 6
```
So, yes, we know that in memory this is a union, but we also know that the data of the union is guaranteed to be properly populated as `Int32` at this point, so just automatically shift the pointer to point at that directly.

I think the solution could be done in two ways. Either recognizing this particular scenario and shifting the pointers (avoids a copy),
or perhaps applying a likely more generally useful solution of rewriting this:
```crystal
if (xx = x)
  ...
```
to this:
```crystal
if x
  xx = x
  ...
```
(which, yes, I originally advertised as a workaround)

**Comment by asterite** [comment]
I'm not sure this is a bug. The way it works in the first example, `xx` is assigned a value that is nilable, and so it makes `xx` be nilable. That happens inside an `if` and so `xx` is known to be not nil to the compiler. However, the variable that represents `xx` is nilable because it was assigned a nilable value. Taking a pointer of it exposes the fact that it's actually nilable (even though the compiler knows that it's not nil at that point, the memory layout of the variable holds the possibility of `nil`).

Given that the workaround is pretty easy, I don't think we should try to fix this, or at least it has a very low priority.

**Comment by oprypin** [comment]
@asterite my last/previous message acknowledges all of that and goes a bit further.

The workaround *looks* easy once you know it, but certainly isn't easy to arrive at (and I worry that the *wrong* workaround is much easier to arrive at).

It's also computationally not the most efficient way to do this.

Sure, probably low priority anyway.
